lib/logstash/outputs/elasticsearch/http_client.rb

```diff
@@ -125,14 +156,85 @@ def close
     # * `:path` - String. The leading path for prefixing Elasticsearch
     #   requests. This is sometimes used if you are proxying Elasticsearch access
     #   through a special http path, such as using mod_rewrite.
-    def build_pool(options)
-      hosts = options[:hosts] || ["127.0.0.1"]
-      client_settings = options[:client_settings] || {}
-      timeout = options[:timeout] || 0
+    def calculate_property(uris, property, default, sniff_check)
+      values = uris.map(&property).uniq
+
+      if sniff_check && values.size > 1
+          raise LogStash::ConfigurationError, "Cannot have multiple values for #{property} in hosts when sniffing is enabled!"
+      end
+
+      uri_value = values.first
+
+      default = nil if default == '' # Blanks are as good as nil
+      uri_value = nil if uri_value == ''
+
+      if default && uri_value && (default != uri_value)
+        raise LogStash::ConfigurationError, "Explicit value for '#{property}' was declared, but it is different in one of the URLs given! Please make sure your URLs are inline with explicit values. The URLs have the property set to '#{uri_value}', but it was also set to '#{default}' explicitly"
+      end
+
+      uri_value || default
+    end
+
+    def sniffing
+      @options[:sniffing]
+    end
+
+    def user
+      calculate_property(uris, :user, @options[:user], sniffing)
+    end
+
+    def password
+      calculate_property(uris, :password, @options[:password], sniffing)
+    end
+
+    def path
+      calculated = calculate_property(uris, :path, client_settings[:path], sniffing)
+      calculated = "/#{calculated}" if calculated && !calculated.start_with?("/")
+      calculated
+    end
+
+    def scheme
+      explicit_scheme = if ssl_options && ssl_options.has_key?(:enabled)
+                          ssl_options[:enabled] ? 'https' : 'http'
+                        else
+                          nil
+                        end
+      
+      calculated_scheme = calculate_property(uris, :scheme, explicit_scheme, sniffing)
+
+      if calculated_scheme && calculated_scheme !~ /https?/
+        raise LogStash::ConfigurationError, "Bad scheme '#{calculated_scheme}' found should be one of http/https"
+      end
+
+      if calculated_scheme && explicit_scheme && calculated_scheme != explicit_scheme
+        raise LogStash::ConfigurationError, "SSL option was explicitly set to #{ssl_options[:enabled]} but a URL was also declared with a scheme of '#{explicit_scheme}'. Please reconcile this"
+      end
+
+      calculated_scheme # May be nil if explicit_scheme is nil!
+    end
+
+    def port
+      # Users can't set this, so we always set the default to nil,
+      # We have an or because we always want to return something
+      calculate_property(uris, :port, nil, sniffing) || 9200
+    end
+    
+    DEFAULT_URIS = [::LogStash::Util::SafeURI.new("http://127.0.0.1")]
+    def uris
+      @options[:hosts] || DEFAULT_URIS
```

Comment:
I'll have to check, but I believe this makes testing much much easier. Let me confirm before I say that for sure.